btree/btree.c

```diff
@@ -0,0 +1,13037 @@
+
+	if (addr != NULL) {
+		nxt  = bnode_get(op, NULL, addr, nxt);
+		if (op->no_op.o_sm.sm_rc < 0)
+			return op->no_op.o_sm.sm_rc;
+		node = op->no_node;
+		m0_rwlock_write_lock(&list_lock);
+		tree = node->n_tree;
+
+		if (tree == NULL) {
+			tree = m0_alloc(sizeof *tree);
+			m0_rwlock_init(&tree->t_lock);
+			m0_rwlock_write_lock(&tree->t_lock);
+
+			tree->t_ref = 1;
+			tree->t_root = node;
+			tree->t_height = bnode_level(node) + 1;
+			bnode_fid(node, &tree->t_fid);
+			bnode_lock(node);
+			node->n_tree = tree;
+			bnode_unlock(node);
+		} else {
+			m0_rwlock_write_lock(&tree->t_lock);
+			tree->t_ref++;
+		}
+
+		op->no_node = tree->t_root;
+		op->no_tree = tree;
+		m0_rwlock_write_unlock(&tree->t_lock);
+		m0_rwlock_write_unlock(&list_lock);
+	} else
+		return M0_ERR(-EINVAL);
+
+	return nxt;
+}
+
+
+/**
+ * Returns the tree to the free tree pool if the reference count for this tree
+ * reaches zero.
+ *
+ * @param tree points to the tree to be released.
+ *
+ * @return Next state to proceed in.
+ */
+static void tree_put(struct td *tree)
+{
+	m0_rwlock_write_lock(&list_lock);
+	m0_rwlock_write_lock(&tree->t_lock);
+	M0_ASSERT(tree->t_ref > 0);
+
+	tree->t_ref--;
+
+	if (tree->t_ref == 0) {
+		m0_rwlock_write_unlock(&tree->t_lock);
+		m0_rwlock_fini(&tree->t_lock);
+		m0_rwlock_write_unlock(&list_lock);
+		m0_free0(&tree);
+		return;
+	}
+	m0_rwlock_write_unlock(&tree->t_lock);
+	m0_rwlock_write_unlock(&list_lock);
+
+}
+
+/**
+ * This function loads the node descriptor for the node at segaddr in memory.
+ * If a node descriptor pointing to this node is already loaded in memory then
+ * this function will increment the reference count in the node descriptor
+ * before returning it to the caller.
+ * If the parameter tree is NULL then the function assumes the node at segaddr
+ * to be the root node and will also load the tree descriptor in memory for
+ * this root node.
+ *
+ * @param op load operation to perform.
+ * @param tree pointer to tree whose node is to be loaded or NULL if tree has
+ *             not been loaded.
+ * @param addr node address in the segment.
+ * @param nxt state to return on successful completion
+ *
+ * @return next state
+ */
+static int64_t bnode_get(struct node_op *op, struct td *tree,
+			 struct segaddr *addr, int nxt)
+{
+	const struct node_type *nt;
+	struct nd              *node;
+	bool                    in_lrulist;
+	uint32_t                ntype;
+
+	/**
+	 * TODO: Include function bnode_access() during async mode of btree
+	 * operations to ensure that the node data is loaded form the segment.
+	 * Also consider adding a state here to return as we might need some
+	 * time to load the node if it is not loaded.
+	 */
+
+	/**
+	 * TODO: Adding list_lock to protect from multiple threads
+	 * accessing the same node descriptor concurrently.
+	 * Replace it with a different global lock once hash
+	 * functionality is implemented.
+	 */
+
+	m0_rwlock_write_lock(&list_lock);
+
+	/**
```

Comment:
Do we need to init at the start of this function op->no_op.o_sm.sm_rc ?

Reply:
No, sm_op_init() gets called before this which will initialize sm_rc to 0 before starting the operation.